added guide for using Microsoft dev-tunnels (#175)
openiddict/openiddict-documentation: guides/getting-started/using-microsoft-dev-tunnels.md

# Using dev tunnels <Badge type="warning" text="client" /><Badge type="danger" text="server" />

> When you are developing an Asp.NET or Asp.NET Core based web application running on your developer machine, usually you can connect to the server using the pre configured `applicationUrl` (e.g https://localhost:44359)

Reviewer: ```suggestion
When you are developing an ASP.NET or ASP.NET Core based web application running on your developer machine, usually you can connect to the server using the pre-configured `applicationUrl` (e.g https://localhost:44359/).
```

If you are, however, integrating an Android or iOS application using MAUI or Avalonia, the app will run either on a physical device or on a local emulator/simulator. In either case using localhost is not an option.

> Instead you can use [Microsoft dev tunnels](https://learn.microsoft.com/en-us/azure/developer/dev-tunnels/overview).

Reviewer: ```suggestion
If you are, however, integrating an Android or iOS application using MAUI or Avalonia, the app will run either on a physical device or on a local emulator/simulator.
In either case using localhost is not an option: instead you can use [Microsoft dev tunnels](https://learn.microsoft.com/en-us/azure/developer/dev-tunnels/overview).
```

This tools provides you with a URL that you can connect to. Any traffic received will be routed to your local web application.

## Microsoft dev tunnels

> Microsoft dev tunnels are available for [Visual Studio Code](https://code.visualstudio.com/docs/editor/port-forwarding) and [Visual Studio](https://learn.microsoft.com/en-us/aspnet/core/test/dev-tunnels?view=aspnetcore-8.0)

Reviewer: ```suggestion
Microsoft dev tunnels are available for [Visual Studio Code](https://code.visualstudio.com/docs/editor/port-forwarding) and [Visual Studio](https://learn.microsoft.com/en-us/aspnet/core/test/dev-tunnels?view=aspnetcore-8.0).
```

> Explaining these tools in detail is outside the scope of this guide. Please refer to the following links to
> - [Create a dev tunnel in Visual Studio Code](https://code.visualstudio.com/docs/editor/port-forwarding)
> - [Create a dev tunnel in Visual Studio 2022](https://learn.microsoft.com/en-us/aspnet/core/test/dev-tunnels?view=aspnetcore-8.0)

Reviewer: ```suggestion
Explaining these tools in detail is outside the scope of this guide. Please refer to the following links to:
- [Create a dev tunnel in Visual Studio Code](https://code.visualstudio.com/docs/editor/port-forwarding).
- [Create a dev tunnel in Visual Studio 2022](https://learn.microsoft.com/en-us/aspnet/core/test/dev-tunnels).
```

> ### Configure the IssuerUrl

Reviewer: ```suggestion
### Configure the issuer
```

> In order to use this tunnel now, you need to set the `IssuerUrl` in the client project (e.g. `MauiProgram.cs`)

Reviewer: Note: the property is named `Issuer`, not `IssuerUrl` (BTW, as a general rule, the more specific "URL" name is not used in OAuth 2.0/OIDC, that both prevent the more generic "URI" concept).

> In order to use this tunnel now, you need to set the `IssuerUrl` in the client project (e.g. `MauiProgram.cs`)

Reviewer: ```suggestion
In order to use this tunnel now, you need to change the `Issuer` in the client project (e.g. `MauiProgram.cs`).

```

    .AddClient(options =>
    {
        // ...

        // Add a client registration matching the client application definition in the server project.
        options.AddRegistration(new OpenIddictClientRegistration
        {
            Issuer = new Uri("https://localhost:44395/", UriKind.Absolute), // [!code --]
            Issuer = new Uri("https://vsr1d2gg-44359.euw.devtunnels.ms/", UriKind.Absolute), // [!code ++]
            ProviderName = "Local",

            ClientId = "maui",

            // This sample uses protocol activations with a custom URI scheme to handle callbacks.
            //
            // For more information on how to construct private-use URI schemes,
            // read https://www.rfc-editor.org/rfc/rfc8252#section-7.1 and
            // https://www.rfc-editor.org/rfc/rfc7595#section-3.8.
            PostLogoutRedirectUri = new Uri("com.openiddict.sandbox.maui.client:/callback/logout/local", UriKind.Absolute),
            RedirectUri = new Uri("com.openiddict.sandbox.maui.client:/callback/login/local", UriKind.Absolute),

            Scopes = { Scopes.Email, Scopes.Profile, Scopes.OfflineAccess, "demo_api" }
        });

        // ...

```

> 

Reviewer: ```suggestion
```

> [!WARNING] Use tunnel domain
>
> One pitfall with dev tunnels is, that they do not forward their domain name to the server but redirect traffic to `localhost`. This means, that even though you connect to your server using the dev tunnels URL (e.g. `https://vsr1d2gg-44359.euw.devtunnels.ms`), the request URL on your server will be `https://localhost:44359`. 
>

> > This causes a variety of issues such as
> > - some OpenIddict WebProviders not working since the cookies provided by them in order to keep necessary state are lost since they are created for the tunnel domain and thus not served to `localhost`
> > - the OpenIddict client stack not being able to connect to your local server from a remote machine (e.g. Android emulator) since OpenId Connect Discovery fetches the `.well-known/openid-configuration` which will have all endpoints configured to `https://localhost:44359` instead of your dev tunnels URL.

Reviewer: ```suggestion
> This causes a variety of issues such as:
> - The request forgery protection feature not working correctly due to the antiforgery cookie created using the tunnel domain (and thus not served to `localhost`) being lost in the process.
>
> - The OpenIddict client stack not being able to connect to the server from a remote machine (e.g. Android emulator) due to the server metadata (`/.well-known/openid-configuration`) containing endpoints starting with `https://localhost:44359/` instead of the dev tunnel URL.
```

>
> To fix this, you **must** check the _Use Tunnel Domain_ checkbox in the _Manage dev tunnel_ dialog 
> ![dev tunnels](using-microsoft-dev-tunnels/devtunnels.png)

> ### Add binding for IIS Express 

Reviewer: ```suggestion
### Add a binding for IIS Express

```

When running your app in IIS Express, you will have edit the _applicationhost.config_ and add a binding for your dev tunnel.
Otherwise you will get an error message when trying to connect to your app via the tunnels URL.

> For example, if your web application listens on `https://localhost:44359` locally and your tunnels URL is `https://vsr1d2gg-44359.euw.devtunnels.ms` you can either add an explicit binding for your URL

Reviewer: ```suggestion
For example, if your web application listens on `https://localhost:44359` locally and your tunnels URL is `https://vsr1d2gg-44359.euw.devtunnels.ms`:
```

```xml{3}
    <bindings>
        <binding protocol="http" bindingInformation="*:55946:localhost" />
        <binding protocol="https" bindingInformation="*:44359:vsr1d2gg-44359.euw.devtunnels.ms" />
        <binding protocol="https" bindingInformation="*:44359:localhost" />
    </bindings>
```

> or add a wildcard binding for that port
> ```xml{3}
>     <bindings>
>         <binding protocol="http" bindingInformation="*:55946:localhost" />
>         <binding protocol="https" bindingInformation="*:44359:*" />
>         <binding protocol="https" bindingInformation="*:44359:localhost" />
>     </bindings>
> ```

Reviewer: Since it's considered bad practice in most cases, let's remove that:

```suggestion
```
